在koa中有两个错误处理：`application.js`中的`onerror`方法和`context.js`中的`onerror`方法。

`application中的onerror方法`：

绑定在koa实例对象上，其监听的是`整个应用的error事件`。

`context中的onerror方法：`

绑定在中间件函数数组生成的Promise的catch中与res对象的onFinished函数的回调的，其就是为了处理`请求or响应`中出现的`error事件`的。

<br/>

### application中的onerror

~~~js
onerror(err) {
  // 判断err是否是 new Error过来的实例
  if (!(err instanceof Error)) throw new TypeError(util.format('non-error thrown: %j', err));
  // 忽略404错误
  if (404 == err.status || err.expose) return;
  // 如果有静默设置也忽略
  if (this.silent) return;
  // 打印错误，定位问题
  const msg = err.stack || err.toString();
  console.error();
  console.error(msg.replace(/^/gm, '  '));
  console.error();
}
~~~
<br/>

### context中的onerror

~~~js
onerror(err) {
  // don't do anything if there is no error.
  // this allows you to pass `this.onerror`
  // to node-style callbacks.
  if (null == err) return;
  // 将错误转化为 Error 实例
  if (!(err instanceof Error)) err = new Error(util.format('non-error thrown: %j', err));

  let headerSent = false;
  if (this.headerSent || !this.writable) {
    headerSent = err.headerSent = true;
  }

  // delegate
  this.app.emit('error', err, this);

  // nothing we can do here other
  // than delegate to the app-level
  // handler and log.
  if (headerSent) {
    return;
  }

  const { res } = this;

  // first unset all headers
  /* istanbul ignore else */
  if (typeof res.getHeaderNames === 'function') {
    res.getHeaderNames().forEach(name => res.removeHeader(name));
  } else {
    res._headers = {}; // Node < 7.7
  }

  // then set those specified
  this.set(err.headers);

  // force text/plain
  this.type = 'text';

  // ENOENT support
  if ('ENOENT' == err.code) err.status = 404;

  // default to 500
  if ('number' != typeof err.status || !statuses[err.status]) err.status = 500;

  // respond
  const code = statuses[err.status];
  const msg = err.expose ? err.message : code;
  this.status = err.status;
  this.length = Buffer.byteLength(msg);
  // res的错误处理
  this.res.end(msg);
}
~~~